File: src/actuator.erl
%%%-------------------------------------------------------------------
%%% @author Borja Esteban
%%% @doc
%%%
%%% @end
%%%-------------------------------------------------------------------
-module(actuator).

%% API
-export([load/1]).
-export_type([id/0, actuator/0, group/0]).

-type id() :: {Name :: atom(), Type :: actuator | group}.
-define(ACTUATOR_ID(Name), {Name,       actuator}).
-define(   GROUP_ID(Name), {Name, actuator_group}).

-record(actuator, {
    id       :: id(),
    function :: {Module :: module(), Name :: atom()}
}).
-type actuator() :: #actuator{}.

-record(group, {
    id        :: id(),
    actuators :: {Module :: module(), Name :: atom()}
}).
-type group() :: #group{}.



%%%===================================================================
%%% API
%%%===================================================================

%%--------------------------------------------------------------------
%% @doc Loads the actuators from the indicated modules. 
%% @end
%%-------------------------------------------------------------------
-spec load(Modules :: [Module :: module()]) -> ok.
load(Modules) -> 
    Sets = set_modules(Modules, sets:new()), 
    edb:write(sets:to_list(Sets)),
    ok.


%%%===================================================================
%%% Internal functions
%%%===================================================================

% --------------------------------------------------------------------
set_modules([Module | Modules], Sets) ->
    set_modules(Modules,
        add_groups(Module:groups(), Module, Sets));
set_modules([], Sets) -> 
    Sets.

add_groups([{Name, Actuators} | Groups], Module, Sets) -> 
    Group = #group{
        id        = ?GROUP_ID(Name), 
        actuators = [?ACTUATOR_ID(N) || N <-Actuators]
    },
    sets:add_element(Group, 
            add_groups(Groups, Module, 
                add_actuators(Actuators, Module, Sets)));
add_groups([], _, Sets) -> 
    Sets.

add_actuators([Name | Actuators], Module, Sets) -> 
    Actuator = #actuator{
        id       = ?ACTUATOR_ID(Name),
        function = {Module, Name}
    },
    add_actuators(Actuators, Module,
        sets:add_element(Actuator, Sets));
add_actuators([], _, Sets) -> 
    Sets.
    

%%====================================================================
%% Eunit white box tests
%%====================================================================

% --------------------------------------------------------------------
% TESTS DESCRIPTIONS -------------------------------------------------

% --------------------------------------------------------------------
% SPECIFIC SETUP FUNCTIONS -------------------------------------------

% --------------------------------------------------------------------
% ACTUAL TESTS -------------------------------------------------------

% --------------------------------------------------------------------
% SPECIFIC HELPER FUNCTIONS ------------------------------------------

File: lib/actuator.ex
defmodule Actuator do
  @moduledoc """
  """
  defstruct function: nil, evolution: []

  @type data() :: map()
  @type score() :: number()
  @type error() :: number()
  @type reason() :: term()
  @type result() ::
          {:ok, score(), data()}
          | {:ok, error(), score(), data()}
          | {:stop, reason()}
          | {:stop, reason(), score()}

  ### =================================================================
  ###  API
  ### =================================================================

  @doc """
  Returns the fields of an actuator structure
  """
  @spec fields() :: [atom()]
  def fields(), do: Map.from_struct(%Actuator{}) |> Map.keys()

  @doc """
  Loads the actuators from the indicated modules.
  Should run inside :mnesia transaction (or Database.run/1)
  """
  @spec load(atom | [atom]) :: any
  def load(modules) when is_list(modules) do
    for m <- modules, do: load(m)
  end

  def load(module) do
    groups = Group.load(module)
    actuators = :lists.append(for g <- groups, do: Group.members(g))

    for name <- Enum.uniq(actuators) do
      Database.new(:actuator, name, new(name, module))
    end
  end

  @doc """
  Performs a random evaluation and if true, mutates the actuator
  """
  @spec mutate(atom) :: atom
  def mutate(name) do
    actuator = Database.read!({:actuator, name})

    case :ltools.rand_scale(actuator.evolution) do
      {} -> name
      new -> new
    end
  end

  @doc """
  Executes an actuator function
  """
  @spec run(%Actuator{}, number(), data()) :: result()
  def run(%Actuator{function: function}, signal, data) do
    apply(function, [signal, data])
  end

  ### =================================================================
  ###  Internal functions
  ### =================================================================

  # Creates an actuator with the defined id and function ------------
  defp new(function_name, module) do
    %Actuator{
      function: Function.capture(module, function_name, 2),
      evolution: apply(module, function_name, [])
    }
  end
end
